xds/internal/xdsclient/tlscreds/bundle_ext_test.go

```diff
@@ -0,0 +1,205 @@
+			wantErrPrefix: "failed to unmarshal config:"},
+		{
+			name:          "Missing private key",
+			jd:            `{"certificate_file":"bar"}`,
+			wantErrPrefix: "pemfile: private key file and identity cert file should be both specified or not specified",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			msg := json.RawMessage(test.jd)
+			if _, err := tlscreds.NewBundle(msg); err == nil || !strings.HasPrefix(err.Error(), test.wantErrPrefix) {
+				t.Errorf("NewBundle(%s): got error %s, want an error with prefix %s", msg, err, test.wantErrPrefix)
+			}
+		})
+	}
+}
+
+func TestCaReloading(t *testing.T) {
+	serverCa, err := os.ReadFile(testdata.Path("x509/server_ca_cert.pem"))
+	if err != nil {
+		t.Fatalf("Failed to read test CA cert: %s", err)
+	}
+
+	// Write CA certs to a temporary file so that we can modify it later.
+	caPath := t.TempDir() + "/ca.pem"
+	err = os.WriteFile(caPath, serverCa, 0644)
+	if err != nil {
+		t.Fatalf("Failed to write test CA cert: %v", err)
+	}
+	cfg := fmt.Sprintf(`{
+		"ca_certificate_file": "%s",
+		"refresh_interval": ".01s"
+	}`, caPath)
+	tlsBundle, err := tlscreds.NewBundle([]byte(cfg))
+	if err != nil {
+		t.Fatalf("Failed to create TLS bundle: %v", err)
+	}
+
+	serverCredentials := grpc.Creds(e2e.CreateServerTLSCredentials(t, tls.NoClientCert))
+	server := stubserver.StartTestService(t, nil, serverCredentials)
+
+	conn, err := grpc.Dial(
+		server.Address,
+		grpc.WithCredentialsBundle(tlsBundle),
+		grpc.WithAuthority("x.test.example.com"),
+	)
+	if err != nil {
+		t.Fatalf("Error dialing: %v", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+	_, err = client.EmptyCall(context.Background(), &testpb.Empty{})
+	if err != nil {
+		t.Errorf("Error calling EmptyCall: %v", err)
+	}
+	// close the server and create a new one to force client to do a new
+	// handshake.
+	server.Stop()
+
+	invalidCa, err := os.ReadFile(testdata.Path("ca.pem"))
+	if err != nil {
+		t.Fatalf("Failed to read test CA cert: %v", err)
+	}
+	// unload root cert
+	err = os.WriteFile(caPath, invalidCa, 0644)
+	if err != nil {
+		t.Fatalf("Failed to write test CA cert: %v", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	for ; ctx.Err() == nil; <-time.After(10 * time.Millisecond) {
+		ss := stubserver.StubServer{
+			Address:    server.Address,
+			EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) { return &testpb.Empty{}, nil },
+		}
+		server = stubserver.StartTestService(t, &ss, serverCredentials)
+
+		// Client handshake should eventually fail because the client CA was
+		// reloaded, and thus the server cert is signed by an unknown CA.
+		t.Log(server)
+		_, err = client.EmptyCall(ctx, &testpb.Empty{})
+		const wantErr = "certificate signed by unknown authority"
+		if status.Code(err) == codes.Unavailable && strings.Contains(err.Error(), wantErr) {
+			// Certs have reloaded.
```

Comment:
In this success case, we don't stop the server. Is this not causing the `leakcheck` to fail?

Reply:
OK, thank you for pointing this out. If I understand correctly, `leakcheck` only runs on tests that use `grpctest`, but only a subset of tests actually use this package. Notably xds/internal/xdsclient/bootstrap/bootstrap_test.go doesn't use it, which is the tests I started from, that's why I missed this pattern, I guess.

So I tried to make all my tests use `grpctest`, but the problem is that `NewBundle` inherently leaks: it creates a file watcher provider that spawns a goroutine, but credential bundles are not closable. I considered making NewBundle return another interface that wraps `credentials.Bundle` with an additional Close method that closes to close the underlying provider. I put that in another PR https://github.com/atollena/grpc-go/pull/3 because I'm not sure we need to go that route. The fact that this bundle cannot be closed is not a big deal in itself since it presumably runs for the lifecycle of the application, but I imagine it's pretty common for bundles to require resource cleanup.

I fixed this specific leak anyway and another one I found along the way where I forgot to call `defer conn.Close()`.

Reply:
Instead of adding a new interface that allows closing, how about if we modify `NewBundle` to return a close func:

```golang
func NewBundle(jd json.RawMessage) (credentials.Bundle, func(), error) { ... }
```

Then, callsites would become:
```golang
bundle, close, err := tlscreds.NewBundle(jd)
if err != nil { ... }
defer close()
```

While the above option is easier to use and results in simpler code than adding a new interface, I feel that the whole business of closing a credentials is rather foreign to us and does not make a whole lot of sense in the real world, where the credentials object would live for as long as the associated ClientConn lives (which is usually the whole lifetime of the application).

@dfawley : Do you have any thoughts on this?

Reply:
Had a discussion with @dfawley and we think the following approach might be the best one to take:

- Since these credentials are created during the xDS bootstrap process as part of the `xdsClient` creation, it makes sense to ensure that things are cleaned up when the `xdsClient` is closed.
- The [ServerConfig](https://github.com/grpc/grpc-go/blob/e20d5ce8c48ee70cfabf6d7dc73c4367ee529e04/xds/internal/xdsclient/bootstrap/bootstrap.go#L129) type should store the actual bundle as an unexported field, instead of storing it as a dial option (or maybe it should store both of them). 
- The `xdsClient` close method defined [here](https://github.com/grpc/grpc-go/blob/e20d5ce8c48ee70cfabf6d7dc73c4367ee529e04/xds/internal/xdsclient/clientimpl.go#L72) should reach into the server config [here](https://github.com/grpc/grpc-go/blob/e20d5ce8c48ee70cfabf6d7dc73c4367ee529e04/xds/internal/xdsclient/bootstrap/bootstrap.go#L326) and call the `Close` method on the newly added field for the bundle, if one is defined. https://github.com/grpc/grpc-go/blob/e20d5ce8c48ee70cfabf6d7dc73c4367ee529e04/clientconn.go#L273 is an example where we check for the implementation of a function (not defined as part of the interface) on a type.

And as far as tests go, they should defer a call to the `Close` method.

@atollena : Let me know if this sounds reasonable. Thanks.

Reply:
I implemented this. The cast to `Close` in the xdsClient is not very satisfying, but it works.

Reply:
> The cast to Close in the xdsClient is not very satisfying

Would you rather prefer the approach of defining an interface in `xds/bootstrap` that embeds `credentials.Bundle` and adds a `Close` method? That would get rid of the type assertion. But it would require a no-op close for the existing credentials implementations, but that should be fine. I'm OK either ways. So, if you strongly prefer that approach, go for it.

Sorry for the back and forth on this.

Reply:
> Sorry for the back and forth on this.

No worries, let's try to get this right.

I wound back and thought about why credentials don't have a `Close` method in the first place. The reason it's OK to not have `Close` method to any of the credentials is that they're built outside the gRPC client that uses them. So there's no need for gRPC itself to ever close them, it is the responsibility of the caller. So the existing interface is flexible enough for all use case.

So by this logic, since we have a factory (the `bootstrap.Credentials` type with the `Build` method), I don't think `Close` belongs in the bundle itself, but in the factory.

So, I think the xDS client should have the ability to invoke additional cleanup steps after it stops. Those steps would be created during client setup. Here are the changes that would be needed to do this:

1. In https://github.com/grpc/grpc-go/blob/bb0d32f07806ec84fc7a7bfeff73a5c3b53555bb/xds/bootstrap/bootstrap.go#L39-L44 Build has an additional `func()` return value. It clutters the interface a bit (and would be a breaking change, I just realized package `xds/bootstrap` is public).
2. Add a set of cleanup function in `ServerConfig` (or make it special for creds for now) that must be invoked when the server is closed, and fill it when the credential bundle is created.
3. Invoke this cleanup function when closing the xDS client at https://github.com/grpc/grpc-go/blob/1505f65518984a2feee728f808d3d1de52783f47/xds/internal/xdsclient/clientimpl.go#L72

I've tried this, you can have a look at https://github.com/atollena/grpc-go/pull/4/files. The main drawback is that we now have this "generic" cleanup field on server config that is only used by credential bundles.

(it's a bit weird imo that the credentials are built inside json unmarshalling, but it is tangential to this issue)

I'm happy to merge this change as-is. I'm also happy to do what I propose above by merging https://github.com/atollena/grpc-go/pull/4 into this one. Please let me know if any of you have a preference.

Reply:
Thanks @atollena 

We are good with this approach. Could of minor comments though:
- There are `ServerConfig`s inside of the `Authority` as well, and those need to be cleaned up as well here: https://github.com/atollena/grpc-go/pull/4/files#diff-af2fba77483e91878bd522d2368b9b17dd00608b11c224e07067ec7f14407a52L88
  - As part of A71 (xDS client fallback that is coming up soon), we will need to support multiple top-level and per-authority `ServerConfig`s and support fallback from higher priority to lower priority. So as part of that, we will need to be able to support closing a connection to a server and recreating it. Currently, the connection to a server is open for the lifetime of the xdsclient.
    - When this happens, we would probably have to change the factory to have a `ParseConfig` and a `Build` method. That way, we can ensure that config parsing is done only at bootstrap time, and everytime we need to create a connection to the server, we will call `Build` on the credentials and when we have to close the connection to the server, we will call the returned close method. This cleanup can be done later on though.
- Some of the tests can be reorganized slightly (using `t.Fatal` instead of `t.Error`) and thereby deferring the call to `stop` instead of calling it in an `else` case

>  I just realized package `xds/bootstrap` is public
  
We have been wanting to make this package internal for a while, but never got to it. It is marked experimental, so breaking changes should be OK for now. But at some point in the near future, we should move it to `xds/internal/...`

Reply:
> There are ServerConfigs inside of the Authority as well,

Yes, that was clearly a bug in my last attempt. Calling cleanups in `authority.Close()` should be enough, since `clientImpl.Close()` doesn't have a clientconn itself, but instead delegates to at least one authority? (https://github.com/grpc/grpc-go/blob/master/xds/internal/xdsclient/clientimpl_authority.go#L48C1-L59)

> When this happens, we would probably have to change the factory to have a ParseConfig and a Build method.

OK, I'll leave it as is for now.

> Some of the tests can be reorganized slightly (using t.Fatal instead of t.Error) and thereby deferring the call to stop instead of calling it in an else case

I've done this for some of the tests, but others can continue with other cases by using Errorf and having the else case stop, so to follow https://google.github.io/styleguide/go/decisions#keep-going I kept those as they were.

> We have been wanting to make this package internal for a while

Since we have a breaking change, this may be an opportunity to make it private? I have no way to gauge how much usage there may be.

Reply:
> Since we have a breaking change, this may be an opportunity to make it private? I have no way to gauge how much usage there may be.

There is one usage inside of Google. But we don't expect much usage outside for now. Even if there was usage, adding a no-op `Close` method should be trivial.

We can move change the package in a follow-up PR, so that we can limit the changes here to actually the mTLS implementation. Thanks for the offer :)